### PR TITLE
Allocate less while constructing cranelift-fuzzgen tests

### PR DESCRIPTION
**Pull request by jameysharp**
These are some steps I was taking toward a bigger refactoring of cranelift-fuzzgen, but this is as far as I got finished before other things came up. I think these are still improvements even without the bigger changes I wanted to make, so I'd like to get them merged.

I believe this PR should not change the input format for cranelift-fuzzgen or cranelift-icache and so should be fine to merge at any time.

cc: @afonso360

edit: I believe this will have merge conflicts with #4824. I'd probably prefer to merge that first and then fix this up rather than the other way around.

### CODE REVIEW

**Review by afonso360 — APPROVED** [review]
LGTM! Thanks!

**Review by fitzgen — APPROVED** [review]
LGTM!

**Comment by jameysharp** [comment]
After resolving merge conflicts I got another 200k runs of this target in locally while waiting for CI to finish, so this seems solid.

**Pull request merged**
